catalog/app/containers/Admin/Settings.tsx

```diff
@@ -0,0 +1,257 @@
+
+  const cancel = React.useCallback(() => {
+    setEditing(false)
+  }, [])
+
+  const handleExited = React.useCallback(() => {
+    // reset the form
+    setFormKey(R.inc)
+  }, [])
+
+  const remove = React.useCallback(async () => {
+    if (editing || removing || !settings?.customNavLink) return
+    // XXX: implement custom MUI Dialog-based confirm?
+    // eslint-disable-next-line no-restricted-globals, no-alert
+    if (!confirm('You are about to remove custom link')) return
+    setRemoving(true)
+    try {
+      await writeSettings(R.dissoc('customNavLink', settings))
+    } catch (e) {
+      // eslint-disable-next-line no-console
+      console.warn('Error saving settings:')
+      // eslint-disable-next-line no-console
+      console.error(e)
+      push("Couldn't save settings, see console for details")
+    } finally {
+      setRemoving(false)
+    }
+  }, [editing, removing, settings, writeSettings, push])
+
+  const onSubmit = React.useCallback(
+    async (values: { url: string; label: string }) => {
+      try {
+        await writeSettings({
+          ...settings,
+          customNavLink: { url: values.url, label: values.label },
+        })
+        setEditing(false)
+        return undefined
+      } catch (e) {
+        // eslint-disable-next-line no-console
+        console.warn('Error saving settings:')
+        // eslint-disable-next-line no-console
+        console.error(e)
+        return { [FF.FORM_ERROR]: "Couldn't save settings, see console for details" }
+      }
+    },
+    [settings, writeSettings],
+  )
+
+  return (
+    <>
+      {settings?.customNavLink ? (
+        <div className={classes.section}>
+          <div className={classes.fields}>
+            <div className={classes.field}>
+              <div className={classes.fieldName}>URL:</div>
+              <div className={classes.fieldValue}>{settings.customNavLink.url}</div>
+            </div>
+            <div className={classes.field}>
+              <div className={classes.fieldName}>Label:</div>
+              <div className={classes.fieldValue}>{settings.customNavLink.label}</div>
+            </div>
+          </div>
+          <M.Button
+            variant="outlined"
+            color="primary"
+            size="small"
+            onClick={edit}
+            disabled={removing}
+          >
+            Edit
+          </M.Button>
+          <M.Box pl={1} />
+          <M.Button color="primary" size="small" onClick={remove} disabled={removing}>
+            Remove
+          </M.Button>
+          {removing && <M.CircularProgress size={24} className={classes.progress} />}
+        </div>
+      ) : (
+        <div className={classes.section}>
+          <div className={classes.notConfigured}>Not configured</div>
+          <M.Button variant="outlined" color="primary" size="small" onClick={edit}>
+            Add link
+          </M.Button>
+        </div>
+      )}
+      <M.Dialog open={editing} onExited={handleExited} fullWidth>
```

Comment:
I'd rather want to see this component divided into two pieces: one for viewing and the second for editing

Reply:
Yes. "View"-part is simple, you could declutter it from state management:

```tsx
const NavLinkEdit = () => { state && <Form />}
const NavLinkAdd = () => <button />
const NavLinkView = () => <div />

const NavLinkEditor = () => (
  {noLink ? <NavLinkAdd /> : <NavLinkView />}
  <NavLinkEdit />
)
```

Reply:
well, it's all coupled quite tight currently, and i dont see a simple way of splitting it clearly without leaking abstractions, passing a ton of callbacks and state around